README
NAME
    Palm::Keyring - Handler for Palm Keyring databases.

DESCRIPTION
    The Keyring PDB handler is a helper class for the Palm::PDB package. It
    parses Keyring for Palm OS databases. See
    <http://gnukeyring.sourceforge.net/>.

    It has the standard Palm::PDB methods with 2 additional public methods.
    Decrypt and Encrypt.

    It currently supports the v4 Keyring databases. The v5 databases from
    the pre-release keyring-2.0 are not supported.

    This module doesn't store the decrypted content. It only keeps it until
    it returns it to you or encrypts it.

SYNOPSIS
        use Palm::PDB;
        use Palm::Keyring;
    
        my $pass = 'password';
        my $file = 'Keys-Gtkr.pdb';
        my $pdb  = new Palm::PDB;
        $pdb->Load($file);
    
        foreach (0..$#{ $pdb->{'records'} }) {
            next if $_ = 0; # skip the password record
            my $rec  = $pdb->{'records'}->[$_];
            my $acct = $pdb->Decrypt($rec, $pass);
            print $rec->{'name'}, ' - ', $acct->{'account'}, "\n";
        }

SUBROUTINES/METHODS
  new
        $pdb = new Palm::Keyring([$password]);

    Create a new PDB, initialized with the various Palm::Keyring fields and
    an empty record list.

    Use this method if you're creating a Keyring PDB from scratch otherwise
    you can just use Palm::PDB::new() before calling Load().

    If you pass in a password, it will initalize the first record with the
    encrypted password.

  Encrypt
        $pdb->Encrypt($rec, $acct[, $password]);

    Encrypts an account into a record, either with the password previously
    used, or with a password that is passed.

    $rec is a record from $pdb->{'records'} or a new_Record(). $acct is a
    hashref in the format below.

        my $acct = {
            name       => $rec->{'name'},
            account    => $account,
            password   => $password,
            notes      => $notes,
            lastchange => {
                year  => 107, # years since 1900
                month =>   0, # 0-11, 0 = January, 11 = December
                day   =>  30, # 1-31, same as localtime
            },
        };

    If you have changed anything other than the lastchange, or don't pass in
    a lastchange key, Encrypt() will generate a new lastchange date for you.

    If you pass in a lastchange field that is different than the one in the
    record, it will honor what you passed in.

    Encrypt() only uses the $acct->{'name'} if there is not already a
    $rec->{'name'}.

  Decrypt
        my $acct = $pdb->Decrypt($rec[, $password]);

    Decrypts the record and returns a hashref for the account as described
    under Encrypt().

        foreach (0..$#{ $pdb->{'records'}) {
            next if $_ == 0;
            my $rec = $pdb->{'records'}->[$_];
            my $acct = $pdb->Decrypt($rec[, $password]);
            # do something with $acct
        }

  Password
        $pdb->Password([$password[, $new_password]]);

    Either sets the password to be used to crypt, or if you pass
    $new_password, changes the password on the database.

    If you have created a new $pdb, and you didn't set a password when you
    called new(), you only need to pass one password and it will set that as
    the password.

    If nothing is passed, it forgets the password that it was remembering.

DEPENDENCIES
    Palm::StdAppInfo

    Digest::MD5

    Crypt::DES

    Readonly

THANKS
    I would like to thank the helpful Perlmonk shigetsu who gave me some
    great advice and helped me get my first module posted.
    <http://perlmonks.org/?node_id=596998>

    I would also like to thank Johan Vromans <jvromans@squirrel.nl> --
    <http://www.squirrel.nl/people/jvromans>. He had his own Palm::KeyRing
    module that he posted a couple of days before mine was ready and he was
    kind enough to let me have the namespace as well as giving me some very
    helpful hints about doing a few things that I was unsure of. He is
    really great.

BUGS AND LIMITATIONS
    Please report any bugs or feature requests to "bug-palm-keyring at
    rt.cpan.org", or through the web interface at <http://rt.cpan.org>. I
    will be notified, and then you'll automatically be notified of progress
    on your bug as I make changes.

AUTHOR
    Andrew Fresh <andrew@mad-techies.org>

LICENSE AND COPYRIGHT
    Copyright 2004, 2005, 2006, 2007 Andrew Fresh, All Rights Reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    Palm::PDB(3)

    Palm::StdAppInfo(3)

    The Keyring for Palm OS website: <http://gnukeyring.sourceforge.net/>

    Johan Vromans also has a wxkeyring app that now uses this module,
    available from his website at <http://www.vromans.org/johan/software/>



CHANGES
0.92

	Fix a problem with the 0.91 API.

0.91	Jan 30 2007, 14:53

	Updated release, now supports multiple records with the same
	name.

	Added support for the "lasst modified" info.

	WARNING: Incompatible API change.

0.90	Jan 28 2007, 21:46

	Updated release, using Palm::PDB. Now supports categories.

	WARNING: Incompatible API change.

0.01	Jan 27 2007, 19:16

	First release, using PDA::Pilot.


Build.PL
#!/usr/bin/perl
# $RedRiver: Build.PL,v 1.6 2007/02/03 00:54:01 andrew Exp $
########################################################################
# Build.PL *** the Build.pl for Palm::Keyring
# 
# 2007.01.28 #*#*# andrew fresh <andrew@mad-techies.org>
########################################################################
use strict;
use warnings;

use Module::Build;
my $build = Module::Build->new
(
	module_name => 'Palm::Keyring',
	license  => 'perl',
	requires => {
		'perl'             => '5.6.1',
		'Palm::StdAppInfo' => '0',
		'Digest::MD5'      => '0',
		'Crypt::DES'       => '0',
	},
	create_makefile_pl => 'traditional',
	create_readme => 1,
	create_html   => 1,
	auto_features => {
		YAML_support =>
		{
			description => "Use YAML.pm to write META.yml files",
			requires => { YAML => ' >= 0.35, != 0.49_01 ' },
		},
		manpage_support =>
		{
			description => "Create Unix man pages",
			requires => { 'Pod::Man' => 0 },
		},
		HTML_support =>
		{
			description => "Create HTML documentation",
			requires => { 'Pod::Html' => 0 },
		},
	},
);
$build->create_build_script;


lib/Palm/KeyRing.pm
#!/usr/bin/perl

package Palm::KeyRing;

$VERSION = "0.92";

use strict;
use warnings;

use Carp;
use Palm::PDB;
use Palm::Raw;

=head1 NAME

Palm::KeyRing - Interface to GNU KeyRing databases

=head1 SYNOPSIS

  use Palm::KeyRing;

  # Open the database.
  my $db = Palm::KeyRing->new("Keys-Gtkr.pdb");

  # Fetch a decryptor (verifies the password).
  my $decryptor = $db->getDecryptor("Secret PassPhrase");
  die("Incorrect password") unless $decryptor;

  # Fetch record by number (first = 1).
  my $rec = $db->getRecord(1);
  # Or fetch record by name.
  # Note that there can be several records with the same name.
  $rec = $db->getRecordsByName("BankAccount")->[0];
  # $rec->{name} the name.
  # $rec->{category} the category index.
  # $rec->{data} contains the encrypteddata.

  # Decrypt record.
  my ($name, $category, $account, $password, $note, $lastmod) =
    $decryptor->decrypt($rec);

=head1 DESCRIPTION

Palm::KeyRing provides a (currently read-only) interface to the
keyring files as used by the GNU KeyRing tool on Palm handhelds.

Records in the keyring file have 5 fields: name, account, password,
note, and lastmod. All fields except name are encrypted.

The lastmod field, if present, will be returned by the decryptor as an
array reference [ year, month, day ], compatible with the first three
elements returned by localtime.

=head1 METHODS

The following are methods of Palm::KeyRing.

=over 4

=item new

Constructor. Takes one argument: the name of the keyring file. 

=item getRecords

Returns the number of records read. Note that the contents will not be
decrypted.

=item getNames

Returns a list (in scalar context: array reference) of all names used
in the keyring file.

=item getRecord

Takes one integer argument, a record number, and returns the record.
This is a hash ref with at least keys "name, "data" and "category".
Note that "category" is the category index.

If no such record exists, returns undef.

=item getRecordsByName

Takes two arguments, a record name, and a category mask. Returns an
list (or arry ref in scalar context) of the corresponding records. See
getRecord above for the return values.

If no such record exists, returns an empty list (or reference).

The category mask is a bit pattern, where bit 0 = category 0
(Unfiled), bit 1 = category 1, and so on. When the category mask is
omitted, returns the results for all categories.

=item getCategory

Takes one argument, a category index. Returns the category as a string.

=item getCategories

In list context, returns a list of categories. In scalar context
returns an array reference.

=item getDecryptor

Takes one argument, the keyring password. If the password is correct,
it returns a decryptor for this keyring. Otherwise, it returns undef.

The decryptor provides one method: decrypt. This method takes one
argument, the record (as a hash ref). It returns a list (in scalar
context: array reference) of the record name, category_name, account,
password, note, and lastmod; all decrypted.

=head1 REQUIREMENTS

PDA::Pilot

Crypt::DES

=head1 DEMO PROGRAM

Palm::KeyRing comes with a nice GUI based demo program wxkeyring. This
program requires WxPerl 0.15 or later.

=head1 BUGS AND LIMITATIONS

Installs Palm::Raw as the default handler for Palm::PDB. This could
have side effects if the rest of the program also uses Palm::PDB.

Please report any bugs or feature requests to bug-palm-keyring at
rt.cpan.org, or through the web interface at http://rt.cpan.org.

=head1 SEE ALSO

Palm::PDB

The Keyring for Palm OS website: http://gnukeyring.sourceforge.net

=head1 AUTHOR

Johan Vromans E<lt>jv@cpan.orgE<gt>.

=head1 LICENCE AND COPYRIGHT

Copyright 2007 Johan Vromans, All Rights Reserved.

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

sub new {
    my ($pkg, $file, %args) = @_;
    $pkg = ref($pkg) || $pkg;
    my $self = {};

    croak("$file: $!\n") unless -f $file && -r $file;
    $self->{db} = Palm::PDB->new;
    $self->{db}->Load($file);
    $self->{filename} = $file;

    # Load the records.
    my $recno = 0;
    foreach ( @{$self->{db}->{records}} ) {
	if ( $recno == 0 ) {
	    $recno++;
	    next;
	}
	my $r = $_->{data};
	$r =~ /^(.+?)\000/s;
	my $name = $1;
	push(@{$self->{recs}->{$name}}, $recno, $_->{category});
	push(@{$self->{names}}, $name);
	$recno++;
    }

    # Load the categories.
    $self->{categories} =
      [ unpack("xx".("A16"x16), $self->{db}->{appinfo}) ];

    # Return the object.
    bless($self, $pkg);
}

sub getDecryptor {
    my ($self, $passwd) = @_;
    Palm::KeyRing::Decryptor->_new
	($self, $self->{db}->{records}->[0]->{data}, $passwd);
}

sub getRecord {
    my ($self, $recno) = @_;
    return undef if $recno >= scalar(@{$self->{db}->{records}});
    $self->{db}->{records}->[$recno];
}

sub getRecordsByName {
    my ($self, $name, $catmask) = @_;
    $catmask = ~0 unless defined $catmask;
    my @ret = ();
    if ( exists $self->{recs}->{$name} ) {
	my @a = @{$self->{recs}->{$name}};
	while ( @a ) {
	    my $rnr = shift(@a);
	    my $cat = shift(@a);
	    next unless $catmask & (1 << $cat);
	    push(@ret, $self->getRecord($rnr));
	}
    }
    wantarray ? @ret : \@ret;
}

sub getRecords {
    my ($self) = @_;
    scalar(@{$self->{db}->{records}})-1;
}

sub getNames {
    my ($self) = @_;
    my @names = @{$self->{names}};
    wantarray ? @names : \@names;
}

sub getCategory {
    my ($self, $cat) = @_;
    return undef unless $self->{categories}->[$cat];
    return $self->{categories}->[$cat];
}

sub getCategoryByName {
    my ($self, $name) = @_;

    unless ( exists($self->{catmap}) ) {
	# Build reverse mapping.
	for ( my $i = 0; $i < 15; $i++ ) {
	    next unless $self->{categories}->[$i];
	    $self->{catmap}->{$self->{categories}->[$i]} = $i;
	}
    }

    $self->{catmap}->{$name};
}

sub getCategories {
    my ($self) = @_;
    wantarray ? @{$self->{categories}} : $self->{categories};
}

package Palm::KeyRing::Decryptor;

use Crypt::DES;
use Digest::MD5 qw(md5);
use Carp;

sub _new {
    my ($pkg, $super, $keyring0, $passwd) = @_;
    $pkg = ref($pkg) || $pkg;

    my $msg = substr($keyring0,0,4).$passwd."\000" x 64;
    $msg = substr($msg,0,64);	# cut to 64 bytes
    my $digest = md5($msg);
    if ( substr($keyring0,4,length($digest)) eq $digest ) {
	my $key = md5($passwd);
	return bless([$super,
		      Crypt::DES->new(substr($key,0,8)),
		      Crypt::DES->new(substr($key,8,8))], $pkg);
    }
    undef;
}

sub decrypt {
    my ($self, $rec) = @_;

    croak("decrypt requires a hash ref as argument")
      unless UNIVERSAL::isa($rec, 'HASH');
    my $data = $rec->{data};
    my $cat = $self->[0]->getCategory($rec->{category});
    my ($name, $raw) = split(/\000/, $data, 2);
    my $out = "";
    for ( my $j=0; $j<int(length($raw) / 8); $j++) {
	my $to = $self->[1]->decrypt( substr($raw,$j*8,8) );
	my $other = $self->[2]->encrypt($to);
	$to = $self->[1]->decrypt($other);
	$out .= $to;
    }
    my ($acc,$pass,$note,$x) = split(/\000/,$out,4);
    $note =~ s/\n+$// if $note;

    if ( $x && (my $packed_date = unpack("n", $x)) ) {
	my @tm = ((($packed_date & 0xFE00) >> 9) + 4,
		  (($packed_date & 0x01E0) >> 5) - 1,
		  ($packed_date & 0x001F),
		  0, 0, 0);
	$x = \@tm;
    }
    else {
	undef $x;
    }

    wantarray ? ($name,$cat,$acc,$pass,$note,$x) : [$name,$cat,$acc,$pass,$note,$x];
}

# Self-test. This requires the Keys-Gtkr.pdb that is in the tests
# directory.

unless ( caller ) {

    # Open the database.
    my $db = Palm::KeyRing->new("Keys-Gtkr.pdb");

    warn("Number of records = ", $db->getRecords, "\n");

    # Verify the password.
    my $decryptor = $db->getDecryptor("secret");
    die("Incorrect password") unless $decryptor;

    # Fetch first record:
    my $rec = $db->getRecord(1);
    # Or fetch a record by its name:
    $rec = $db->getRecordsByName("Bank",
				 1 << $db->getCategoryByName("Banking")
				)->[0];

    # Decrypt record.
    my ($name, $cat, $account, $password, $note, $x) =
      $decryptor->decode($rec);

    print("Name:     $name\n",
	  "Category: $cat\n",
	  "Account:  $account\n",
	  "Password: $password\n",
	  "Note:     $note\n",
	  "Modified: ", $x ? sprintf("%d-%02d-%02d",
				     1900 + $x->[0],
				     1 + $x->[1],
				     $x->[2]) : "", "\n",
	 );


}

1;


scripts/example3.pl
#!/usr/bin/perl
use strict;
use warnings;
use Getopt::Long;
Getopt::Long::Configure('bundling');
use Term::ReadLine;

use YAML;

use Palm::PDB;
use Palm::Keyring;

my $Default_File = $ENV{'HOME'} . '/.jpilot/Keys-Gtkr.pdb';
my $PDBFile;
my $Categories;
my $Names;
my $Action_List;
my $Action_Show;

my $result = GetOptions (
	"file|f=s"        => \$PDBFile,
	"categories|c:s@" => \$Categories,
	"name|n=s@"       => \$Names,
	"list|l"          => \$Action_List,
	"show|s"          => \$Action_Show,
);

$PDBFile ||= $Default_File;
my $pdb = new Palm::PDB();
$pdb->Load($PDBFile) || die "Couldn't load '$PDBFile': $!";

if ($Action_List) {
	show_list();
} elsif ($Action_Show) {
	show_items($Names);
} elsif (defined $Categories) {
	show_categories();
} else {
	help();
}

exit;


sub show_list
{
	print "Showing List\n";
	foreach (0..$#{ $pdb->{'records'} }) {
        next if $_ == 0;
        my $r = $pdb->{'records'}->[$_];
		my $category = 
			$pdb->{'appinfo'}->{'categories'}->[ $r->{'category'} ]->{'name'};

		my $matched = 0;
		foreach my $cat (@{ $Categories }) {
			$matched++ if uc($category) eq uc($cat);
		}
		foreach my $name (@{ $Names}) {
			$matched++ if uc($r->{'name'}) eq uc($name);
		}
		next if ( @{ $Categories } || @{ $Names } ) && not $matched;

		# XXX Fix up formatting
		print $r->{'name'} .
			":" .
			$r->{'category'} . 
			":" .
			$category .
			"\n";
			
	}
}

sub show_categories
{
	foreach my $c (sort @{ $pdb->{'appinfo'}->{'categories'} }) {
		next unless $c->{'name'};
		# Fix formatting
		print $c->{'name'}, "\n";
	}
}

sub show_items
{
	get_password() || die "Couldn't decrypt file!";

	foreach (0..$#{ $pdb->{'records'} }) {
        next if $_ == 0;
        my $r = $pdb->{'records'}->[$_];

		my $category = 
			$pdb->{'appinfo'}->{'categories'}->[ $r->{'category'} ]->{'name'};

		my $matched = 0;
		foreach my $cat (@{ $Categories }) {
			$matched++ if uc($category) eq uc($cat);
		}
		foreach my $name (@{ $Names}) {
			$matched++ if uc($r->{'name'}) eq uc($name);
		}
		next if ( @{ $Categories } || @{ $Names } ) && not $matched;

        my $a = $pdb->Decrypt($r);

		# XXX Fix up formatting
		print $r->{'name'} .  "\n\t" .
			"Category: " . $category .  "\n\t" .
			"Account:  " . $a->{'account'} .  "\n\t" .
			"Password: " . $a->{'password'} .  "\n";
			print "\tNotes: " . $a->{'notes'} . "\n" if $a->{'notes'};
	}

}

sub add_item
{
	die "not implemented!";
}

sub delete_item
{
	die "not implemented!";
}

sub get_password
{
	while (1) {
		print "Enter Password: ";

		system("stty", "-echo");
		chop(my $read = <STDIN>);
        system("stty", "echo");
        print "\n";

		$read =~ s/^\s*//;
		$read =~ s/\s*$//;

		#return 1 if
		$pdb->Password($read) && return 1;
		#print Dump $read, $pdb;
		#exit;
	}
	return undef;
}


sub help
{
	print STDERR "$0 [options] action\n";
	exit 255;
}


lib/Palm/Keyring.pm
package Palm::Keyring;

# $RedRiver: Keyring.pm,v 1.25 2007/02/03 01:12:21 andrew Exp $
#
# Perl class for dealing with Keyring for Palm OS databases.
#
#   This started as Memo.pm, I just made it work for Keyring.

use strict;
use warnings;
use Carp;

use base qw/ Palm::StdAppInfo /;

use Digest::MD5 qw(md5);
use Crypt::DES;

my $ENCRYPT    = 1;
my $DECRYPT    = 0;
my $MD5_CBLOCK = 64;
my $kSalt_Size = 4;
my $EMPTY      = q{};
my $SPACE      = q{ };
my $NULL       = chr 0;

our $VERSION = 0.94;

sub new {
    my $classname = shift;
    my $pass      = shift;

    # Create a generic PDB. No need to rebless it, though.
    my $self = $classname->SUPER::new(@_);

    $self->{'name'}    = 'Keys-Gtkr';    # Default
    $self->{'creator'} = 'Gtkr';
    $self->{'type'}    = 'Gkyr';

    # The PDB is not a resource database by
    # default, but it's worth emphasizing,
    # since MemoDB is explicitly not a PRC.
    $self->{'attributes'}{'resource'} = 0;

    # Initialize the AppInfo block
    $self->{'appinfo'} = {};

    # Add the standard AppInfo block stuff
    Palm::StdAppInfo::seed_StdAppInfo( $self->{'appinfo'} );

    # Set the version
    $self->{'version'} = 4;

    if ( defined $pass ) {
        $self->Password($pass);
    }

    return $self;
}

sub import {
    Palm::PDB::RegisterPDBHandlers( __PACKAGE__, [ 'Gtkr', 'Gkyr' ], );
    return 1;
}

sub ParseRecord {
    my $self     = shift;

    my $rec = $self->SUPER::ParseRecord(@_);

    # skip the 0 record that holds the password
    return $rec if ! exists $self->{'records'}; 
    return $rec if ! exists $rec->{'data'};

    my ( $name, $encrypted ) = split /$NULL/xm, $rec->{'data'}, 2;

    return $rec if ! $encrypted;
    delete $rec->{'data'};
    $rec->{'name'} = $name;
    $rec->{'encrypted'} = $encrypted;

    return $rec;
}

sub PackRecord {
    my $self = shift;
    my $rec  = shift;

    if ($rec->{'encrypted'}) {
        if (! defined $rec->{'name'}) {
            $rec->{'name'} = $EMPTY;
        }
        $rec->{'data'} = join $NULL, $rec->{'name'}, $rec->{'encrypted'};
        delete $rec->{'name'};
        delete $rec->{'encrypted'};
    }

    return $self->SUPER::PackRecord($rec, @_);
}

sub Encrypt {
    my $self = shift;
    my $rec  = shift;
    my $data = shift;
    my $pass = shift || $self->{'password'};

    if ( ! $pass) {
        croak("'password' not set!\n");
    }

    if ( ! $rec) {
        croak("Needed parameter 'record' not passed!\n");
    }

    if ( ! $data) {
        croak("Needed parameter 'data' not passed!\n");
    }

    if ( ! $self->Password($pass)) {
        croak("Incorrect Password!\n");
    }

    $self->{'digest'}   ||= _calc_keys( $pass );

    $data->{'account'}  ||= $EMPTY;
    $data->{'password'} ||= $EMPTY;
    $data->{'notes'}    ||= $EMPTY;

    my $changed      = 0;
    my $need_newdate = 0;
    my $acct = {};
    if ($rec->{'encrypted'}) {
        $acct = $self->Decrypt($rec, $pass);
        foreach my $key (keys %{ $data }) {
            next if $key eq 'lastchange';
            if ($data->{$key} ne $acct->{$key}) {
                $changed = 1;
                last;
            }
        }
        if ( exists $data->{'lastchange'} && exists $acct->{'lastchange'} && (
            $data->{'lastchange'}->{day}   != $acct->{'lastchange'}->{day}   ||
            $data->{'lastchange'}->{month} != $acct->{'lastchange'}->{month} ||
            $data->{'lastchange'}->{year}  != $acct->{'lastchange'}->{year}
        )) {
            $changed = 1;
            $need_newdate = 0;
        } else {
            $need_newdate = 1;
        }

    } else {
        $changed = 1;
    }

    # no need to re-encrypt if it has not changed.
    return 1 if ! $changed;

    my ($day, $month, $year);

    if ($data->{'lastchange'} && ! $need_newdate ) {
        $day   = $data->{'lastchange'}->{'day'}   || 1;
        $month = $data->{'lastchange'}->{'month'} || 0;
        $year  = $data->{'lastchange'}->{'year'}  || 0;

        # XXX Need to actually validate the above information somehow
        if ($year >= 1900) {
            $year -= 1900;
        }
    } else {
        $need_newdate = 1;
    }

    if ($need_newdate) {
        ($day, $month, $year) = (localtime)[3,4,5];
    }
    $year -= 4;
    $month++;


    my $p = $day | ($month << 5) | ($year << 9);
    my $packeddate = pack 'n', $p;

    my $plaintext = join $NULL, 
        $data->{'account'}, $data->{'password'}, $data->{'notes'}, $packeddate;

    my $encrypted = _crypt3des( $plaintext, $self->{'digest'}, $ENCRYPT );

    return if ! $encrypted;

    $rec->{'attributes'}{'Dirty'} = 1;
    $rec->{'attributes'}{'dirty'} = 1;
    $rec->{'name'}    ||= $data->{'name'};
    $rec->{'encrypted'} = $encrypted;

    return 1;
}

sub Decrypt {
    my $self = shift;
    my $rec  = shift;
    my $pass = shift || $self->{'password'};

    if ( ! $pass) {
        croak("'password' not set!\n");
    }

    if ( ! $rec) {
        croak("Needed parameter 'record' not passed!\n");
    }

    if ( ! $self->Password($pass)) {
        croak("Invalid Password!\n");
    }

    if ( ! $rec->{'encrypted'} ) {
        croak("No encrypted content!");
    }

    $self->{'digest'} ||= _calc_keys( $pass );

    my $decrypted = 
        _crypt3des( $rec->{'encrypted'}, $self->{'digest'}, $DECRYPT );
    my ( $account, $password, $notes, $packeddate ) = split /$NULL/xm,
          $decrypted, 4;

    my %Modified;
    if ($packeddate) {
        my $u = unpack 'n', $packeddate;
        my $year  = (($u & 0xFE00) >> 9) + 4; # since 1900
        my $month = (($u & 0x01E0) >> 5) - 1; # 0-11
        my $day   = (($u & 0x001F) >> 0);     # 1-31

        %Modified = (
            year   => $year,
            month  => $month || 0,
            day    => $day   || 1,
        );
    }

    return {
        name       => $rec->{'name'},
        account    => $account,
        password   => $password,
        notes      => $notes,
        lastchange => \%Modified,
    };
}

sub Password {
    my $self = shift;
    my $pass = shift;
    my $new_pass = shift;

    if (! $pass) {
        delete $self->{password};
	return 1;
    }

    if (! exists $self->{'records'}) {
        # Give the PDB the first record that will hold the encrypted password
        $self->{'records'} = [ $self->new_Record ];

        return $self->_password_update($pass);
    }

    if ($new_pass) {
        my @accts = ();
        foreach my $i (0..$#{ $self->{'records'} }) {
            if ($i == 0) {
                push @accts, undef;
                next;
            }
            my $acct = $self->Decrypt($self->{'records'}->[$i], $pass);
            if ( ! $acct ) {
                croak("Couldn't decrypt $self->{'records'}->[$i]->{'name'}");
            }
            push @accts, $acct;
        }

        if ( ! $self->_password_update($new_pass)) {
            croak("Couldn't set new password!");
        }
        $pass = $new_pass;

        foreach my $i (0..$#accts) {
            next if $i == 0;
            delete $self->{'records'}->[$i]->{'encrypted'};
            $self->Encrypt($self->{'records'}->[$i], $accts[$i], $pass);
        }
    }

    return $self->_password_verify($pass);
}

sub _calc_keys {
    my $pass = shift;
    if (! defined $pass) { croak('No password defined!'); };

    my $digest = md5($pass);

    my ( $key1, $key2 ) = unpack 'a8a8', $digest;

    #--------------------------------------------------
    # print "key1: $key1: ", length $key1, "\n";
    # print "key2: $key2: ", length $key2, "\n";
    #--------------------------------------------------

    $digest = unpack 'H*', $key1 . $key2 . $key1;

    #--------------------------------------------------
    # print "Digest: ", $digest, "\n";
    # print length $digest, "\n";
    #--------------------------------------------------

    return $digest;
}

sub _password_verify {
    my $self = shift;
    my $pass = shift;

    if (! $pass) { croak('No password specified!'); };

    if (defined $self->{'password'} && $pass eq $self->{'password'}) {
        # already verified this password
        return 1;
    }

    # AFAIK the thing we use to test the password is
    #     always in the first entry
    my $data = $self->{'records'}->[0]->{'data'};

    #die "No encrypted password in file!" unless defined $data;
    if ( ! defined $data) { return; };

    $data =~ s/$NULL$//xm;

    my $salt = substr $data, 0, $kSalt_Size;

    my $msg = $salt . $pass;

    $msg .= "\0" x ( $MD5_CBLOCK - length $msg );

    my $digest = md5($msg);

    if ( $data eq $salt . $digest ) {

# May as well generate the keys we need now, since we know the password is right
        $self->{'digest'} = _calc_keys($pass);
        if ( $self->{'digest'} ) {
            $self->{'password'} = $pass;
            return 1;
        }
    }
    return;
}

sub _password_update {

    # It is very important to Encrypt after calling this
    #     (Although it is generally only called by Encrypt)
    # because otherwise the data will be out of sync with the
    # password, and that would suck!
    my $self = shift;
    my $pass = shift;

    if (! defined $pass) { croak('No password specified!'); };

    my $salt;
    for ( 1 .. $kSalt_Size ) {
        $salt .= chr int rand 255;
    }

    my $msg = $salt . $pass;

    $msg .= "\0" x ( $MD5_CBLOCK - length $msg );

    my $digest = md5($msg);

    my $data = $salt . $digest;    # . "\0";

    # AFAIK the thing we use to test the password is
    #     always in the first entry
    $self->{'records'}->[0]->{'data'} = $data;

    $self->{'password'} = $pass;
    $self->{'digest'}   = _calc_keys( $self->{'password'} );

    return 1;
}

sub _crypt3des {
    my ( $plaintext, $passphrase, $flag ) = @_;

    $passphrase   .= $SPACE x ( 16 * 3 );
    my $cyphertext = $EMPTY;

    my $size = length $plaintext;

    #print "STRING: '$plaintext' - Length: " . (length $plaintext) . "\n";

    my @C;
    for ( 0 .. 2 ) {
        $C[$_] =
          new Crypt::DES( pack 'H*', ( substr $passphrase, 16 * $_, 16 ));
    }

    for ( 0 .. ( ($size) / 8 ) ) {
        my $pt = substr $plaintext, $_ * 8, 8;

        #print "PT: '$pt' - Length: " . length($pt) . "\n";
        if (! length $pt) { next; };
        if ( (length $pt) < 8 ) {
            if ($flag == $DECRYPT) { croak('record not 8 byte padded'); };
            my $len = 8 - (length $pt);

            #print "LENGTH: $len\n";
            #print "Binary:    '" . unpack("b*", $pt) . "'\n";
            $pt .= ($NULL x $len);

            #print "PT: '$pt' - Length: " . length($pt) . "\n";
            #print "Binary:    '" . unpack("b*", $pt) . "'\n";
        }
        if ( $flag == $ENCRYPT ) {
            $pt = $C[0]->encrypt($pt);
            $pt = $C[1]->decrypt($pt);
            $pt = $C[2]->encrypt($pt);
        }
        else {
            $pt = $C[0]->decrypt($pt);
            $pt = $C[1]->encrypt($pt);
            $pt = $C[2]->decrypt($pt);
        }

        #print "PT: '$pt' - Length: " . length($pt) . "\n";
        $cyphertext .= $pt;
    }

    $cyphertext =~ s/$NULL+$//xm;

    #print "CT: '$cyphertext' - Length: " . length($cyphertext) . "\n";

    return $cyphertext;
}

1;
__END__

=head1 NAME

Palm::Keyring - Handler for Palm Keyring databases.

=head1 DESCRIPTION

The Keyring PDB handler is a helper class for the Palm::PDB package. It
parses Keyring for Palm OS databases.  See
L<http://gnukeyring.sourceforge.net/>.

It has the standard Palm::PDB methods with 2 additional public methods.  
Decrypt and Encrypt.

It currently supports the v4 Keyring databases.  The v5 databases from 
the pre-release keyring-2.0 are not supported.

This module doesn't store the decrypted content.  It only keeps it until it 
returns it to you or encrypts it.

=head1 SYNOPSIS

    use Palm::PDB;
    use Palm::Keyring;
    
    my $pass = 'password';
    my $file = 'Keys-Gtkr.pdb';
    my $pdb  = new Palm::PDB;
    $pdb->Load($file);
    
    foreach (0..$#{ $pdb->{'records'} }) {
        next if $_ = 0; # skip the password record
        my $rec  = $pdb->{'records'}->[$_];
        my $acct = $pdb->Decrypt($rec, $pass);
        print $rec->{'name'}, ' - ', $acct->{'account'}, "\n";
    }

=head1 SUBROUTINES/METHODS

=head2 new

    $pdb = new Palm::Keyring([$password]);

Create a new PDB, initialized with the various Palm::Keyring fields
and an empty record list.

Use this method if you're creating a Keyring PDB from scratch otherwise you 
can just use Palm::PDB::new() before calling Load().

If you pass in a password, it will initalize the first record with the encrypted 
password.

=head2 Encrypt

    $pdb->Encrypt($rec, $acct[, $password]);

Encrypts an account into a record, either with the password previously 
used, or with a password that is passed.

$rec is a record from $pdb->{'records'} or a new_Record().
$acct is a hashref in the format below.

    my $acct = {
        name       => $rec->{'name'},
        account    => $account,
        password   => $password,
        notes      => $notes,
        lastchange => {
            year  => 107, # years since 1900
            month =>   0, # 0-11, 0 = January, 11 = December
            day   =>  30, # 1-31, same as localtime
        },
    };

If you have changed anything other than the lastchange, or don't pass in a 
lastchange key, Encrypt() will generate a new lastchange date for you.

If you pass in a lastchange field that is different than the one in the
record, it will honor what you passed in.

Encrypt() only uses the $acct->{'name'} if there is not already a $rec->{'name'}.

=head2 Decrypt

    my $acct = $pdb->Decrypt($rec[, $password]);

Decrypts the record and returns a hashref for the account as described 
under Encrypt().  

    foreach (0..$#{ $pdb->{'records'}) {
        next if $_ == 0;
        my $rec = $pdb->{'records'}->[$_];
        my $acct = $pdb->Decrypt($rec[, $password]);
        # do something with $acct
    }

=head2 Password

    $pdb->Password([$password[, $new_password]]);

Either sets the password to be used to crypt, or if you pass $new_password, 
changes the password on the database.

If you have created a new $pdb, and you didn't set a password when you 
called new(), you only need to pass one password and it will set that as 
the password. 

If nothing is passed, it forgets the password that it was remembering.

=head1 DEPENDENCIES

Palm::StdAppInfo

Digest::MD5

Crypt::DES

Readonly

=head1 THANKS

I would like to thank the helpful Perlmonk shigetsu who gave me some great advice
and helped me get my first module posted.  L<http://perlmonks.org/?node_id=596998>

I would also like to thank 
Johan Vromans
E<lt>jvromans@squirrel.nlE<gt> -- 
L<http://www.squirrel.nl/people/jvromans>.
He had his own Palm::KeyRing module that he posted a couple of days before 
mine was ready and he was kind enough to let me have the namespace as well 
as giving me some very helpful hints about doing a few things that I was 
unsure of.  He is really great.

=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests to
C<bug-palm-keyring at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.  I will be notified, and then you'll automatically be
notified of progress on your bug as I make changes.

=head1 AUTHOR

Andrew Fresh E<lt>andrew@mad-techies.orgE<gt>

=head1 LICENSE AND COPYRIGHT

Copyright 2004, 2005, 2006, 2007 Andrew Fresh, All Rights Reserved.

This program is free software; you can redistribute it and/or 
modify it under the same terms as Perl itself.

=head1 SEE ALSO

Palm::PDB(3)

Palm::StdAppInfo(3)

The Keyring for Palm OS website: 
L<http://gnukeyring.sourceforge.net/>

Johan Vromans also has a wxkeyring app that now uses this module, available 
from his website at L<http://www.vromans.org/johan/software/>
